lib/resty/iprepd.lua

```diff
@@ -0,0 +1,76 @@
+local cjson = require('cjson')
+local lrucache = require('resty.lrucache')
+local http = require('resty.http')
+
+function fatal_error(error_msg)
+  ngx.log(ngx.ERR, error_msg)
+  os.exit(1)
+end
+
+local _M = {}
+local mt = { __index = _M }
+
+function _M.new(options)
+  local iprepd_url = options.url or 'http://localhost:8080/'
+  local cache_ttl = options.cache_ttl or 30
+  local iprepd_threshold = options.threshold
+  if not iprepd_threshold then
+    fatal_error('Need to pass in a threshold')
+  end
+  local iprepd_api_key = options.api_key
+  if not iprepd_api_key then
+    fatal_error('Need to pass in an api_key')
+  end
+
+  -- TODO: Make configurable?
+  local cache, err = lrucache.new(200)
+  if not cache then
+    fatal_error('failed to create the cache: ' .. (err or 'unknown'))
+  end
+
+  local self = {
+    url = iprepd_url,
+    threshold = iprepd_threshold,
+    api_key_hdr = {
+      ['Authorization'] = 'APIKey ' .. iprepd_api_key,
+    },
+    cache_ttl = cache_ttl,
+    timeout = options.timeout or 10,
+    cache = cache,
+  }
+  return setmetatable(self, mt)
+end
+
+function _M.check(self, ip)
+  local httpc = http.new()
+  -- set timeout in ms
+  httpc:set_timeout(self.timeout)
+
+  -- Get reputation for ip
+  local resp = self.cache:get(ip)
+  if not resp then
+    resp, err = httpc:request_uri(self.url .. '/' .. ip, {
+      method  = "GET",
+      headers = self.api_key_hdr,
+    })
+    if err then
+      ngx.log(ngx.ERR, 'Error with request to iprepd: ' .. err)
+    else
+      self.cache:set(ngx.var.remote_addr, resp, self.cache_ttl)
```

Comment:
should we extract the reputation out of the JSON here and store just that value in the cache instead of the original JSON string? That seems to be the only value we are using from it so it would save requiring a decode operation after each cache lookup. We could also reject bad JSON here and not store it in the cache, and validate the returned reputation is in range (e.g., >= 0 && <= 100).